R/linear_pool_sample.R

```diff
@@ -42,15 +35,77 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
     )
     weights_col_name <- "weight"
   }
+  weight_by_cols <- colnames(weights)[colnames(weights) != weights_col_name]
   unique_weights <- unique(weights[[weights_col_name]])
 
-  if (length(unique_weights) != 1 || length(unique_provided_samples) != 1 || !is.null(n_output_samples)) {
-    cli::cli_abort(
-      "The requested ensemble calculation doesn't satisfy all conditions:
-      1) {.arg model_out_tbl} contains the same number of samples from each component model,
-      2) {.arg weights} are {.val NULL} or equal for every model,
-      3) {.arg n_output_samples} = {.val NULL}"
-    )
+  if (length(unique_weights) != 1) {
+    cli::cli_abort("{.arg weights} must be {.val NULL} or equal for every model")
+  }
+
+  if (!is.null(n_output_samples)) {
+    samples_per_combo <- model_out_tbl |>
+      dplyr::group_by(dplyr::across(dplyr::all_of(c("model_id", compound_taskid_set)))) |>
+      dplyr::summarize(provided_samples = length(unique(.data[["output_type_id"]]))) |>
+      dplyr::ungroup() |>
+      tidyr::complete(
+        !!!rlang::syms(c("model_id", compound_taskid_set)),
+        fill = list(provided_samples = 0)
+      ) |>
+      dplyr::left_join(weights, weight_by_cols) |>
+      dplyr::mutate(target_samples = ifelse(
+        .data[["provided_samples"]] <= 0,
+        0,
+        floor(.data[[weights_col_name]] * n_output_samples)
+      ))
```

Comment:
It seems like we probably need to re-normalize the model weights within groups defined by `compound_taskid_set`.  Borrowing an example from the last round of review, suppose `compound_taskid_set` includes `"location"`, and we have three equally-weighted models that each submitted 100 samples, but one of those models does not submit predictions for one location.  Then this calculation would result in only 66 ensemble samples.  I think this can be remedied by updating the weights using code something like this (untested):
```suggestion
      dplyr::group_by(dplyr::across(dplyr::all_of(compound_taskid_set))) |>
      dplyr::mutate(
        effective_weight = as.integer(.data[["provided_samples"]] > 0) * .data[[weights_col_name]],
        effective_weight = .data[["effective_weight"]] / sum(.data[["effective_weight"]]),
        target_samples = floor(.data[["effective_weight"]] * n_output_samples)
      ) |>
      dplyr::ungroup()
```